### ISSUE TITLE
Introduce a private Ethereum network and write a single test on it as a PoC

### ISSUE THREAD

**Issue by tiabc** [status-im/status-go#312]
# Problem Overview

We have a number of integration tests under `./geth/api`, `./geth/jail` and `./geth/node` and they primarily run on Ropsten (some on Rinkeby). This is a problem because it takes longer and is more expensive than it should be:
1. We have to spend our faucet balance on tests.
2. This introduces delays until blockchain is in sync to the most recent state, which may sometimes take long.
3. We just clutter the network with no reason.

Also, this is a great chance to create documentation (in markdown, preferably) on how to add new networks.

# Implementation

If one wants to run classic geth on a private chain, they end up writing something like this:
`geth --nodiscover --maxpeers 1 --networkid 118 --datadir "data1" --port 30301 console`

More details on set up may be found here: https://github.com/ethereum/go-ethereum/wiki/Private-network. It should become clear on which options to use in code for a private network.

This doesn't imply that one should start a separate node but rather serves as guidance about which parameters to utilise in the internal node.

Concrete steps to be accomplished while dealing with the issue:
1. Add a new constant similar to: https://github.com/status-im/status-go/blob/0c4603d825ae5c4be3d87e99eff6e849452e25e6/geth/params/defaults.go#L98
2. Add a new test for default config similar to: https://github.com/status-im/status-go/blob/c85525e139745720c67a03d3bcb5635e0687383e/cmd/statusd/utils.go#L176
3. This will also show you how to create a genesis file which should be the input for new chains.

# Acceptance criteria

1. There's a new private chain named `StatusChain` based on a newly created genesis which can be used in tests.
2. `StatusChain` genesis contains enough balance for sending transactions from our pre-created test accounts.
3. A copy of `BackendTestSuite.TestSendEtherTx` passes on `StatusChain` (just copy the test, the next step will be to make the network configurable) with disabled upstream.

# Notes

Please, refer to #220 for more background and details.
That'd also be great if you could provide some random thoughts what you think we can improve everywhere.

# Future steps

These steps are out of scope but are just worth keeping in mind while working on the rest.

1. Add an ability to choose network to run tests on. Also delete a temporarily copied `TestSendEther`.
2. There's an instruction on how to add new networks.
3. Code should be refactored the way that it's clear from the code how to do it (instead of an instruction).

**Comment by divan** [comment]
As `geth` is Go-based (and we, in fact, vendoring it in codebase already), would it be feasible to run the equivalent of `geth --nodiscover --maxpeers 1 --networkid 118 --datadir "data1" --port 30301 console` as a part of the test code directly, instead of launching as separate process/container?

Perhaps it'd be possible also to use memory listener and avoid using OS sockets and network completely to run these tests?

**Comment by screwyprof** [comment]
I started working on this issue and having done a quick research I came to a conclusion that the task is too large to be done at once.

I've been thinking how to do the testing properly and by this time I have the following understanding: 

In order to run tests locally, so that we didn't depend on Ropsten/Rinkenby we need to create a local private network with a custom genesis, with mining running and with a couple of test accounts pre-created.

I suggest splitting this task into subtasks. 
1. Create a private testing network and an example test case which is run on this network, for example, it can be transferring ether from one account to another (TestSendEtherTx)
2. Change the other tests so that they could run on the local testing network

In order for me to implement the first part, I need some help. 

I noted that now it's possible to run tests without having internet access to the testing network in a case when tests were run before, which seems quite strange to me. What I've been meaning to say is that it's possible to re-run tests locally. I started investigating and found out that there are local and upstream transactions, but I wasn't able to understand how they differ from each other. What are upstream and local transactions? Would you be so kind as to elaborate on it, please?

Apart from that, I need to what kind of environment is used to run tests? Can we use Docker for example? If no, then where should we run a testing network?

Is there any documentation or maybe some guides for newbies to start with?

@adambabik, @divan, @tiabc

**Comment by adambabik** [comment]
Hey @screwyprof!

> What are upstream and local transactions?

There is no distinction between upstream and local transactions. Both are regular ethereum transactions but there is a difference how they're created. Transactions sent using the local node go through the same flow as using geth (it's almost the same as you would run `geth console` and execute `web3.sendTransaction({ params })`). Transactions sent using the upstream (remote) node are signed locally by status-go and sent using `eth_sendRawTransaction` RPC call to that remote node. From the user or status-go API perspective there should be no difference.

> Apart from that, I need to what kind of environment is used to run tests? Can we use Docker for example? If no, then where should we run a testing network?

Currently, we use native go tools to run tests. But as you described, the private network requires a few services up and running. Personally, I think Docker is a proper tool to be used. I imagine that I can run the private network using `docker-compose up` command and then `go test ./some/package`.

> Is there any documentation or maybe some guides for newbies to start with?

I am not sure there is more information available at the moment :(

**Comment by screwyprof** [comment]
@adambabik Thank you very much for your response.

Could you please elucidate a little further on tests? I don't really understand what exactly we are testing. For example, if we run a test  build/env.sh go test -timeout 40m -v ./geth/api/... -testify.m ^TestSendEtherTx$

It works correctly even if we don't have internet access. (I ran the test once, disconnected from the internet and ran it again and it still passes.) Isn't it strange that we are testing transactions without having access to some other nodes? What exactly do we test? It seems like some over-testing, because that is exactly what is done by original geth client. In other words, we are testing that ethereum is able to run transactions instead of testing our wrapping API.

**Comment by adambabik** [comment]
> `go test -timeout 40m -v ./geth/api/... -testify.m ^TestSendEtherTx$`

This runs a single test `TestSendEtherTx`, which is an integration test and performs the complete flow of sending a transaction. As I mentioned once, status-go implements a mechanism which queues transactions and sends them only after a confirmation with a password. Also, status-go introduces some patches in go-ethereum that are related to sending transactions.

> It works correctly even if we don't have internet access.

I think it's exactly the same with `geth`. As far as I know, the Internet is not required and the transaction will be sent later when the connection is back.

> It seems like some over-testing, because that is exactly what is done by original geth client. In other words, we are testing that ethereum is able to run transactions instead of testing our wrapping API.

Yes and no. As I mentioned the flow is different in status-go and we have some patches in go-ethereum. I agree that there are too many integration tests, however, some of them are needed as well.

**Comment by screwyprof** [comment]
I managed to run a private network with 2 mining nodes on my local mac and ran a simple test case on it (TestSendEtherTxUpstream). 

Everything is great, the test passes.I used a few hand-crafted bash scripts to init and run them. Now I want to create a few docker containers (miners, bootstrap) and some basic scripts to manage them. It seems to me that it would be enough for this task in order for us to start writing new tests running on a private network in upstream mode.

I think it would be better to create another subtask to investigate how to run those tests that are run on a local status node (the tests that use s.LightEthereumService().StatusBackend) as oppose to ones run with Upsteam to use a private network.

**Comment by divan** [comment]
I think it's fairly easy to build/start docker containers directly from the Go tests. Here is an example implementation of similar tests - a bit outdated perhaps, but it shows the point: https://divan.github.io/posts/integration_testing/

**Comment by screwyprof** [comment]
It can be done, but what for? What can be gained by starting them from code? 
I think it would be better (at least at the beginning) to start the testing network separately and just ping it from tests. We won't have to start/stop containers on every run. If I understand it correctly, the major idea now is to make tests more stable and decrease the testing time.

**Comment by adambabik** [comment]
My experience from running and managing containers directly from the tests is not good (also did that in Go). It seems easy but you end up with quite a lot of boilerplate code that you need to maintain. Also, when you start creating networks, linking containers etc. it gets pretty complicated.

That's why `docker-compose` was created. I would go for it as it's the easiest and it can start multiple linked containers with one command. It's tested, easy to use and everyone is familiar with it.

> We won't have to start/stop containers on every run. If I understand it correctly, the major idea now is to make tests more stable and decrease the testing time.

Correct. I can't think of a scenario that would require to stop and start the network.

**Comment by divan** [comment]
@adambabik agree, but if you need to launch just one or two containers, it's much easier and faster to do it from go code, so default `go test` would work. Once it gets more complicated, docker-compose is necessary, yes.

As for the boilerplate, there are libraries to faciliate that (I used https://github.com/ory/dockertest a lot). But the main point of this solution is to be able to use tests without prior knowledge of scripts needed to be executed before testing, which is nice and much simpler. Also less code to maintain and debug errors if any.
